### PR TITLE
Add Toric Code Rectangle to ReCirq (#1)

### PR DESCRIPTION
**Pull request by dstrain115**
- This PR adds the Toric Code Rectangle class to ReCirq.

### CODE REVIEW

**Comment by mpharrigan** [comment]
is there any way to split this up into a couple of PRs?

**Comment by dstrain115** [comment]
> is there any way to split this up into a couple of PRs?

I can split it up later.  I would suggest starting with the toric_code_rectangle if you only want to review one file at a time.

File: recirq/toric_code/toric_code_rectangle.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+# limitations under the License.
+"""Module for representing, preparing, and analyzing toric code states.
+
+Rectangular patches of the toric code are represented by ToricCodeRectangle. These are used to
+make Cirq circuits to make the toric code ground state with ToricCodeStatePrep. Resulting plaquette
+parity expectation values are stored with ToricCodePlaquettes, which can be visualized using
+ToricCodePlotter.
+"""
+
+from __future__ import annotations
```
Comment (mpharrigan) [inline recirq/toric_code/toric_code_rectangle.py:22]: do we need this?

File: recirq/toric_code/toric_code_rectangle.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        self.qubits_reading_order = self._qubits_reading_order()
+
+    def __repr__(self) -> str:
+        return (
+            f"ToricCodeRectangle(origin_qubit=cirq.GridQubit{self.origin_qubit}, "
+            f"row_vector={tuple(self.row_vector)}, rows={self.rows}, cols={self.cols})"
+        )
+
+    @staticmethod
+    def q_displaced(qubit: cirq.GridQubit, displacement: np.ndarray) -> cirq.GridQubit:
```
Comment (mpharrigan) [inline recirq/toric_code/toric_code_rectangle.py:128]: optional: this can be a free function

**Review by mpharrigan — COMMENTED** [review]
`toric_code_rectangle.py` looks good to me

File: recirq/toric_code/toric_code_rectangle_test.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+import pytest
+
+from . import toric_code_rectangle as tcr
+
+
+def q(i: int, j: int) -> cirq.GridQubit:
+    return cirq.GridQubit(i, j)
+
+
+class TestToricCodeRectangle:
```
Comment (mpharrigan) [inline recirq/toric_code/toric_code_rectangle_test.py:28]: can we change this to normal pytest just-a-bunch-of-free-funcitons. for the setup code, you can define a pytest fixture or (imo even better) just factor it out into a helper function and call it at the beginning of each test

File: recirq/toric_code/toric_code_rectangle.py (hunk 1 above)
Comment (dstrain115) [reply recirq/toric_code/toric_code_rectangle.py:22]: Removed.

File: recirq/toric_code/toric_code_rectangle.py (hunk 2 above)
Comment (dstrain115) [reply recirq/toric_code/toric_code_rectangle.py:128]: Done.

File: recirq/toric_code/toric_code_rectangle_test.py (hunk 3 above)
Comment (dstrain115) [reply recirq/toric_code/toric_code_rectangle_test.py:28]: Done.

**Review by dstrain115 — COMMENTED** [review]
Addressed comments and pared down to just toric_code_rectangle.

**Comment by dstrain115** [comment]
Let me know if you have further comments so I can continue down the PR chain.

**Review by mpharrigan — APPROVED** [review]
LGTM. Can we get a sign-off from @kjsatz

File: recirq/toric_code/toric_code_rectangle.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+            for row, col in self.x_plaquette_indices()
+            for qubit in self.x_plaquette_to_qubits(row, col)
+        }
+
+    def _qubits_reading_order(self) -> Tuple[cirq.GridQubit, ...]:
+        """Qubits listed in "reading order."
+
+        The qubits are enumerated starting with the "captain" of x_plaquette (0, 0)
+        being map[1] and then reading right (the first self.cols qubits are the captains
+        of row 0) and going down row by row.
```
Comment (kjsatz) [inline recirq/toric_code/toric_code_rectangle.py:157]: Update docstring to reflect the tuple and zero-indexing (it looks like the old dict and one-indexing docstring)

**Review by kjsatz — APPROVED** [review]
Thank you!

File: recirq/toric_code/toric_code_rectangle.py (hunk 4 above)
Comment (dstrain115) [reply recirq/toric_code/toric_code_rectangle.py]: Done.

**Pull request merged**